system/include/compat/avxintrin.h

```diff
@@ -11,14 +11,505 @@
+    ((__f32x4)__a)[wasm_i32x4_extract_lane(__c, 3) & 3]);
+}
+
+static __inline __m256 __DEFAULT_FN_ATTRS _mm256_permutevar_ps(__m256 __a,
+                                                               __m256i __c) {
+  __m256 ret;
+  ret.v0 = _mm_permutevar_ps(__a.v0, __c.v0);
+  ret.v1 = _mm_permutevar_ps(__a.v1, __c.v1);
+  return ret;
+}
+
+#define _mm_permute_pd(__a, __imm)                                             \
+  __extension__({                                                              \
+    (__m128d) wasm_i64x2_shuffle(                                              \
+      (__m128d)(__a), (__m128d)(__a), ((__imm) & 1), (((__imm) >> 1) & 1));    \
+  })
+
+#define _mm256_permute_pd(__a, __imm)                                          \
+  __extension__({                                                              \
+    _mm256_set_m128d(_mm_permute_pd((__a).v1, (__imm) >> 2),                   \
+                     _mm_permute_pd((__a).v0, (__imm)));                       \
+  })
+
+#define _mm_permute_ps(__a, __imm)                                             \
+  __extension__({                                                              \
+    (__m128) wasm_i32x4_shuffle((__m128)(__a),                                 \
+                                (__m128)(__a),                                 \
+                                ((__imm) & 3),                                 \
+                                (((__imm) >> 2) & 3),                          \
+                                (((__imm) >> 4) & 3),                          \
+                                (((__imm) >> 6) & 3));                         \
+  })
+
+#define _mm256_permute_ps(__a, __imm)                                          \
+  __extension__({                                                              \
+    _mm256_set_m128(_mm_permute_ps((__a).v1, (__imm)),                         \
+                    _mm_permute_ps((__a).v0, (__imm)));                        \
+  })
+
+static inline __m128d select4d(__m256d __a, __m256d __b, const int imm8) {
+  const int index = imm8 & 0x3;
+  __m128d tmp;
+  switch (index) {
+    case 0:
+      tmp = __a.v0;
+      break;
+    case 1:
+      tmp = __a.v1;
+      break;
+    case 2:
+      tmp = __b.v0;
+      break;
+    case 3:
+      tmp = __b.v1;
+      break;
+  }
+  if (imm8 & 0x8) {
+    tmp = (__m128d)wasm_i64x2_const_splat(0);
+  }
+  return tmp;
+}
+
+static inline __m128 select4(__m256 __a, __m256 __b, const int imm8) {
+  const int index = imm8 & 0x3;
+  __m128 tmp;
+  switch (index) {
+    case 0:
+      tmp = __a.v0;
+      break;
+    case 1:
+      tmp = __a.v1;
+      break;
+    case 2:
+      tmp = __b.v0;
+      break;
+    case 3:
+      tmp = __b.v1;
+      break;
+  }
+  if (imm8 & 0x8) {
+    tmp = (__m128)wasm_i64x2_const_splat(0);
+  }
+  return tmp;
+}
+
+static inline __m128i select4i(__m256i __a, __m256i __b, const int imm8) {
```

Comment:
To avoid pollution, these would probably be good to be named something like `__avx_select4i`, `__avx_select4`, `__avx_select4d` in case a user codebase might have a function named e.g. `select4`.